### PR TITLE
[SYCL][CUDA][PI] Add CUDA-specific pi_mem_advice values

### PR DESCRIPTION
**Pull request by t4c1**
Add CUDA-specific pi_mem_advice values and implement support for them in PI.

Tests: https://github.com/intel/llvm-test-suite/pull/602

### CODE REVIEW

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 1]
```diff
@@ -4709,9 +4709,31 @@ pi_result cuda_piextUSMEnqueueMemAdvise(pi_queue queue, const void *ptr,
...
+    case PI_MEM_ADVISE_CUDA_SET_PREFERRED_LOCATION_HOST:
+    case PI_MEM_ADVISE_CUDA_UNSET_PREFERRED_LOCATION_HOST:
+    case PI_MEM_ADVISE_CUDA_SET_ACCESSED_BY_HOST:
+    case PI_MEM_ADVISE_CUDA_UNSET_ACCESSED_BY_HOST:
+      result = PI_CHECK_ERROR(cuMemAdvise(
+          (CUdeviceptr)ptr, length,
+          (CUmem_advise)(advice -
+                         (PI_MEM_ADVISE_CUDA_SET_PREFERRED_LOCATION_HOST -
+                          PI_MEM_ADVISE_CUDA_SET_PREFERRED_LOCATION)),
+          -1));
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:4732]:
```suggestion
          CU_DEVICE_CPU));
```

File: sycl/include/CL/sycl/detail/pi.h [hunk 2]
```diff
@@ -440,7 +440,17 @@ typedef enum {
...
+  PI_MEM_ADVISE_CUDA_SET_READ_MOSTLY = 1,
+  PI_MEM_ADVISE_CUDA_UNSET_READ_MOSTLY = 2,
+  PI_MEM_ADVISE_CUDA_SET_PREFERRED_LOCATION = 3,
+  PI_MEM_ADVISE_CUDA_UNSET_PREFERRED_LOCATION = 4,
+  PI_MEM_ADVISE_CUDA_SET_ACCESSED_BY = 5,
+  PI_MEM_ADVISE_CUDA_UNSET_ACCESSED_BY = 6,
+  PI_MEM_ADVISE_CUDA_SET_PREFERRED_LOCATION_HOST = 7,
+  PI_MEM_ADVISE_CUDA_UNSET_PREFERRED_LOCATION_HOST = 8,
+  PI_MEM_ADVISE_CUDA_SET_ACCESSED_BY_HOST = 9,
+  PI_MEM_ADVISE_CUDA_UNSET_ACCESSED_BY_HOST = 10,
```
Comment (steffenlarsen) [inline sycl/include/CL/sycl/detail/pi.h:453]: These are all CUDA specific, however most other PI enums are built around the OpenCL counter-part. Granted there are no OpenCL descriptors yet, I would still prefer that the CUDA-specific options are moved to start at a number that isn't 1 as we should leave some buffer for OpenCL descriptors in the future.

**Review by steffenlarsen — CHANGES_REQUESTED** [review]
Nice addition but we should be consistent with other implementations.

**Comment by AerialMantis** [comment]
@steffenlarsen @smaslov-intel could we get another review on this one please.

**Review by steffenlarsen — APPROVED** [review]
LGTM! Sorry for the delay.

File: sycl/include/CL/sycl/detail/pi.h [hunk 3]
```diff
@@ -440,7 +440,17 @@ typedef enum {
 
 typedef enum {
   // Device-specific value opaque in PI API.
-  PI_MEM_ADVISE_UNKNOWN
+  PI_MEM_ADVISE_UNKNOWN,
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:443]: NIT: to follow PI API notation and repeat enum typename:
```suggestion
  PI_MEM_ADVIcE_UNKNOWN,
```

**Review by steffenlarsen — APPROVED** [review]

**Comment by bader** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/602

**Comment by bader** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/602

**Review by bader — APPROVED** [review]

**Comment by bader** [comment]
[2022-02-04T14:18:47.318Z] Failed Tests (2):
[2022-02-04T14:18:47.318Z]   SYCL :: DiscardEvents/discard_events_usm.cpp
[2022-02-04T14:18:47.318Z]   SYCL :: DiscardEvents/discard_events_usm_ooo_queue.cpp

@t4c1, it looks like these tests should be updated as well.

**Comment by t4c1** [comment]
The tests are now updated.

**Comment by bader** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/602

**Pull request merged**